azurerm/resource_arm_iothub_shared_access_policy.go

```diff
@@ -0,0 +1,356 @@
+			if d.IsNewResource() && requireResourcesToBeImported {
+				return tf.ImportAsExistsError("azurerm_iothub_shared_access_policy", resourceId)
+			}
+			accessPolicies = append(accessPolicies, expandedAccessPolicy)
+			alreadyExists = true
+		} else {
+			accessPolicies = append(accessPolicies, existingAccessPolicy)
+		}
+	}
+
+	if d.IsNewResource() {
+		accessPolicies = append(accessPolicies, expandedAccessPolicy)
+	} else if !alreadyExists {
+		return fmt.Errorf("Unable to find Shared Access Policy %q defined for IotHub %q (Resource Group %q)", keyName, iothubName, resourceGroup)
+	}
+
+	iothub.Properties.AuthorizationPolicies = &accessPolicies
+
+	future, err := client.CreateOrUpdate(ctx, resourceGroup, iothubName, iothub, "")
+	if err != nil {
+		return fmt.Errorf("Error updating IotHub %q (Resource Group %q) with Shared Access Profile %q: %+v", iothubName, resourceGroup, keyName, err)
+	}
+
+	if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
+		return fmt.Errorf("Error waiting for IotHub %q (Resource Group %q) to finish updating Shared Access Profile %q: %+v", iothubName, resourceGroup, keyName, err)
+	}
+
+	d.SetId(resourceId)
+
+	return resourceArmIotHubSharedAccessPolicyRead(d, meta)
+}
+
+type accessRights struct {
+	registryRead   bool
+	registryWrite  bool
+	serviceConnect bool
+	deviceConnect  bool
+}
+
+func expandAccessRights(d *schema.ResourceData) string {
+
+	var possibleAccessRights = []struct {
+		schema string
+		right  string
+	}{
+		{"registry_read", "RegistryRead"},
+		{"registry_write", "RegistryWrite"},
+		{"service_connect", "ServiceConnect"},
+		{"device_connect", "DeviceConnect"},
+	}
+	actualRights := make([]string, 0)
+	// iteration order is important here, so we cannot use a map
+	for _, possibleRight := range possibleAccessRights {
+		if d.Get(possibleRight.schema).(bool) {
+			actualRights = append(actualRights, possibleRight.right)
+		}
+	}
+	strRights := strings.Join(actualRights, ", ")
+	return strRights
+}
+
+func flattenAccessRights(r devices.AccessRights) accessRights {
+	rights := accessRights{
+		registryRead:   false,
+		registryWrite:  false,
+		deviceConnect:  false,
+		serviceConnect: false,
+	}
+
+	actualAccessRights := strings.Split(string(r), ",")
+
+	for _, right := range actualAccessRights {
+		switch strings.ToLower(strings.Trim(right, " ")) {
+		case "registrywrite":
+			rights.registryWrite = true
+			// RegistryWrite implies RegistryRead.
+			// What's more, creating a Access Policy with both RegistryRead and RegistryWrite
+			// only really sets RegistryWrite permission, which then also implies RedistryRead
+			fallthrough
+		case "registryread":
+			rights.registryRead = true
+		case "deviceconnect":
+			rights.deviceConnect = true
+		case "serviceconnect":
+			rights.serviceConnect = true
+		}
+	}
+
+	return rights
+}
+
+func resourceArmIotHubSharedAccessPolicyRead(d *schema.ResourceData, meta interface{}) error {
```

Comment:
Could we place this and delete after the CreateUpdate function?

Reply:
done